File: typespec/hub/profilepage.go
package hub

import (
	"time"

	"github.com/vetchium/vetchium/typespec/common"
)

type AddOfficialEmailRequest struct {
	Email common.EmailAddress `json:"email" validate:"required,email"`
}

type VerifyOfficialEmailRequest struct {
	Email common.EmailAddress `json:"email" validate:"required,email"`
	Code  string              `json:"code"  validate:"required"`
}

type TriggerVerificationRequest struct {
	Email common.EmailAddress `json:"email" validate:"required,email"`
}

type DeleteOfficialEmailRequest struct {
	Email common.EmailAddress `json:"email" validate:"required,email"`
}

type OfficialEmail struct {
	Email            common.EmailAddress `json:"email"`
	LastVerifiedAt   *time.Time          `json:"last_verified_at"`
	VerifyInProgress bool                `json:"verify_in_progress"`
}

type GetBioRequest struct {
	Handle string `json:"handle" validate:"required"`
}

type ColleagueConnectionState string

const (
	CanSendRequestCCState         ColleagueConnectionState = "CAN_SEND_REQUEST"
	CannotSendRequestCCState      ColleagueConnectionState = "CANNOT_SEND_REQUEST"
	RequestSentPendingCCState     ColleagueConnectionState = "REQUEST_SENT_PENDING"
	RequestReceivedPendingCCState ColleagueConnectionState = "REQUEST_RECEIVED_PENDING"
	ConnectedCCState              ColleagueConnectionState = "CONNECTED"
	RejectedByMeCCState           ColleagueConnectionState = "REJECTED_BY_ME"
	RejectedByThemCCState         ColleagueConnectionState = "REJECTED_BY_THEM"
	UnlinkedByMeCCState           ColleagueConnectionState = "UNLINKED_BY_ME"
	UnlinkedByThemCCState         ColleagueConnectionState = "UNLINKED_BY_THEM"
)

type Bio struct {
	Handle                   string                   `json:"handle"`
	FullName                 string                   `json:"full_name"`
	ShortBio                 string                   `json:"short_bio"`
	LongBio                  string                   `json:"long_bio"`
	VerifiedMailDomains      []string                 `json:"verified_mail_domains"`
	ColleagueConnectionState ColleagueConnectionState `json:"colleague_connection_state"`
}

type UpdateBioRequest struct {
	FullName *string `json:"full_name" validate:"required,min=1,max=64"`
	ShortBio *string `json:"short_bio" validate:"required,min=1,max=64"`
	LongBio  *string `json:"long_bio"  validate:"required,min=1,max=1024"`
}

type UploadProfilePictureRequest struct {
	Image []byte `json:"image" validate:"required"`
}

File: typespec/hub/profilepage.ts
import { EmailAddress } from "../common/common";

export interface AddOfficialEmailRequest {
  email: EmailAddress;
}

export interface VerifyOfficialEmailRequest {
  email: EmailAddress;
  code: string;
}

export interface TriggerVerificationRequest {
  email: EmailAddress;
}

export interface DeleteOfficialEmailRequest {
  email: EmailAddress;
}

export interface OfficialEmail {
  email: EmailAddress;
  last_verified_at?: string;
  verify_in_progress: boolean;
}

export interface GetBioRequest {
  handle: string;
}

export type ColleagueConnectionState = {
  CAN_SEND_REQUEST: {};
  CANNOT_SEND_REQUEST: {};
  REQUEST_SENT_PENDING: {};
  REQUEST_RECEIVED_PENDING: {};
  CONNECTED: {};
  REJECTED_BY_ME: {};
  REJECTED_BY_THEM: {};
  UNLINKED_BY_ME: {};
  UNLINKED_BY_THEM: {};
}[keyof {
  CAN_SEND_REQUEST: {};
  CANNOT_SEND_REQUEST: {};
  REQUEST_SENT_PENDING: {};
  REQUEST_RECEIVED_PENDING: {};
  CONNECTED: {};
  REJECTED_BY_ME: {};
  REJECTED_BY_THEM: {};
  UNLINKED_BY_ME: {};
  UNLINKED_BY_THEM: {};
}];

export interface Bio {
  handle: string;
  full_name: string;
  short_bio: string;
  long_bio: string;
  verified_mail_domains?: string[];
  colleague_connection_state: ColleagueConnectionState;
}

export interface UpdateBioRequest {
  full_name?: string;
  short_bio?: string;
  long_bio?: string;
}

export interface UploadProfilePictureRequest {
  image: Uint8Array;
}
